### ISSUE TITLE
Login Filter (LDAP like)

### ISSUE THREAD

**Issue by rullzer** [nextcloud/user_saml#179]
It would be good to have the possibility to allow/deny login for (or the absence of) specific shibboleth attributes.

E.g. eduPersonAffiliation: Only allow login if value="employee"

Maybe a more general approach would be better. Maybe a login-condition-query parser for something like this:

urn:oid:zzzzzz=="xxxxxx" AND NOT urn:oid:yyyyy="pppppp"

**Comment by wiswedel** [comment]
@schiessle Do you have any ideas for this? Don't we already have drilled out attribute interpretation with Global Scale?

**Comment by schiessle** [comment]
Shouldn't this be handled  by the IDP. The IDP authenticates the user and should have all the rules about which users are allowed to access which service.

**Comment by wiswedel** [comment]
I'm not a SAML expert (at all), but my customer says they can only authenticate but not authorize via SAML. Isn't that why so many have both SAML _and_ LDAP configured, to apply filters via LDAP which wouldn't be available via SAML IDP?

**Comment by blizzz** [comment]
> Isn't that why so many have both SAML and LDAP configured, to apply filters via LDAP which wouldn't be available via SAML IDP?

I'd say so, too. Configure LDAP with all the constraints, require the user to be existing on a user backend, and there you go.

**Comment by wiswedel** [comment]
There are a lot of customers out there who don't have LDAP additionally to their SAML. That's what this is mostly for.

**Comment by tuxmainy** [comment]
Hi,
> Shouldn't this be handled by the IDP?
Nope, the IDP authenticates you and gives informations (like groups) to the SP. The SP itself has to decide.

I think handling this by groups would be the easiest way. The admin should set a/multiple group(s) per provider in NC and user_saml should check the groups given by the IDP against the groups configured in NC:

match -> access granted
no match -> access denied

I tried to hack this into the plugin but the NC plugin architecture isn't the easiest of all, eh?

regards
Daniel

**Comment by Superjamp** [comment]
Implementing a regex as a deny (and/or allow) filter for just the uid is quite simple.  
But I see we have very different needs for this implementation.

**Comment by CarlSchwan** [comment]
> Shouldn't this be handled by the IDP. The IDP authenticates the user and should have all the rules about which users are allowed to access which service.

Agree, at least with Authentik but I suppose other IDP, I can allow a specific group/user to access an application . I find this a bit cleaner than restricting on the SAML side

**Comment by blizzz** [comment]
> > Shouldn't this be handled by the IDP. The IDP authenticates the user and should have all the rules about which users are allowed to access which service.
> 
> Agree, at least with Authentik but I suppose other IDP, I can allow a specific group/user to access an application . I find this a bit cleaner than restricting on the SAML side

This actually is the job of the IdP, to grant access to the service provider only to those who are configured to get it. But you can already do this when using LDAP in combination, as pointed out at https://github.com/nextcloud/user_saml/issues/179#issuecomment-604916546.  

This should not be implemented here, really.

**Comment by tuxmainy** [comment]
> This actually is the job of the IdP, to grant access to the service provider only to those who are configured to get it. But you can already do this when using LDAP in combination, as pointed out at [#179 (comment)](https://github.com/nextcloud/user_saml/issues/179#issuecomment-604916546).

Did I understood you correctly? Admins having an IdP/SAML should also install a LDAP server, somehow sync the users from the IdP into the LDAP server so they can use the filter which is implemented in LDAP? Sounds weird to me ;)

> This should not be implemented here, really.

Then this should be removed from LDAP. Why handling an IdP/SAML different than LDAP?

That said, I think the discussion is about the thin line between authentication and authorization:

1. no matter what, the IdP authenticates the user, we all agree here
2. but is giving access to a service (i.e. nextcloud) part of authentication or authorization?

If it's part of authentication everything is fine. The nextcloud administrator agreed to "externalize" the authentication via SAML. But if you think it's part of authorization you just lost control because you cannot filter. It's like sharing a file in nextcloud but the admin is fully in control (not just restricting) of who you share this file with.

On the other hand: This issue was opened 2018. Maybe we should sacrifice this issue to the god of honesty and close it :D

regards
Daniel

**Comment by blizzz** [comment]
> Did I understood you correctly? Admins having an IdP/SAML should also install a LDAP server, somehow sync the users from the IdP into the LDAP server so they can use the filter which is implemented in LDAP? Sounds weird to me ;)

No, they typically (but not necessarily) have an LDAP server, that is feeding their IdP.

They should confiugre the SP in their IdP properly and only allow the authorized people to sign in for that service.

Configuring the LDAP backend would be a work around if for whatever reason the above is not possible.



>>    This should not be implemented here, really.

> Then this should be removed from LDAP. Why handling an IdP/SAML different than LDAP?

For bananas aren't oranges.

LDAP is a user directly and may hand out a lot more information to a connected client, and is also queryable.

SAML (or ODIC) for that matter only provide bare necessary information upon an authorization request. The client/service otherwise knows nothing, and also cannot query the SAML server.

> no matter what, the IdP authenticates the user, we all agree here
but is giving access to a service (i.e. nextcloud) part of authentication or authorization?

Yes, by definition, SAML (and also OIDC) is also responsible for authorization.

> SAML consists of building-block components that, when put together, allow a number of use cases to be supported. The components primarily permit transfer of identity, authentication, attribute, and authorization information between autonomous organizations that have an established trust relationship.

[source](https://docs.oasis-open.org/security/saml/Post2.0/sstc-saml-tech-overview-2.0.html) ff


> On the other hand: This issue was opened 2018. Maybe we should sacrifice this issue to the god of honesty and close it :D

🤝
